### PR TITLE
iox-1394 fix axivion warnings for multiple files part 2

### PR DESCRIPTION
**Pull request by elBoberido**
## Pre-Review Checklist for the PR Author

1. [x] Code follows the coding style of [CONTRIBUTING.md][contributing]
1. [x] Tests follow the [best practice for testing][testing]
1. [x] Changelog updated [in the unreleased section][changelog] including API breaking changes
1. [x] Branch follows the naming format (`iox-123-this-is-a-branch`)
1. [x] Commits messages are according to this [guideline][commit-guidelines]
    - [x] Commit messages have the issue ID (`iox-#123 commit text`)
    - [x] Commit author matches [Eclipse Contributor Agreement][eca] (and ECA is signed)
1. [x] Update the PR title
   - Follow the same conventions as for commit messages
   - Link to the relevant issue
1. [x] Relevant issues are linked
1. [x] Add sensible notes for the reviewer
1. [x] All checks have passed (except `task-list-completed`)
1. [x] All touched (C/C++) source code files from `iceoryx_hoofs` are added to `./clang-tidy-diff-scans.txt`
1. [x] Assign PR to reviewer

[commit-guidelines]: https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html
[eca]: http://www.eclipse.org/legal/ECA.php
[contributing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/CONTRIBUTING.md#coding-style
[testing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/concepts/best-practice-for-testing.md
[changelog]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/release-notes/iceoryx-unreleased.md

## Checklist for the PR Reviewer

- [x] Commits are properly organized and messages are according to the guideline
- [x] Code according to our coding style and naming conventions
- [x] Unit tests have been written for new behavior
- [x] Public API changes are documented via doxygen
- [x] Copyright owner are updated in the changed files
- [x] All touched (C/C++) source code files from `iceoryx_hoofs` have been added to `./clang-tidy-diff-scans.txt`
- [x] PR title describes the changes

## Post-review Checklist for the PR Author

1. [x] All open points are addressed and tracked via issues

## References

- Relates to #1394

### CODE REVIEW

File: iceoryx_hoofs/functional/include/iox/detail/storable_function.inl [hunk 1]
```diff
@@ -309,7 +317,9 @@ template <uint64_t Capacity, typename ReturnType, typename... Args>
 template <typename T>
 inline constexpr uint64_t storable_function<Capacity, signature<ReturnType, Args...>>::required_storage_size() noexcept
 {
-    return sizeof(T) + alignof(T) - 1;
+    const auto size{sizeof(T)};
+    const auto alignment{alignof(T)};
+    return size + alignment - 1;
```
Comment (FerdinandSpitzschnueffler) [inline iceoryx_hoofs/functional/include/iox/detail/storable_function.inl:322]: This helps to suppress the warning about missing parentheses? Oh boy...

File: iceoryx_hoofs/vocabulary/include/iox/detail/expected.inl [hunk 2]
```diff
@@ -36,6 +36,7 @@ inline success<T>::success(T&& t) noexcept
 {
 }
 
+// AXIVION Next Construct AutosarC++19_03-A15.4.2 : Intentional behavior. 'success' is not intended to be used with a type which throws
```
Comment (FerdinandSpitzschnueffler) [inline iceoryx_hoofs/vocabulary/include/iox/detail/expected.inl:39]: `FaultDetection-NoexceptViolations` must also be  suppressed.

File: iceoryx_hoofs/vocabulary/include/iox/detail/expected.inl [hunk 3]
```diff
@@ -222,11 +224,12 @@ inline const ValueType& expected<ValueType, ErrorType>::operator*() const noexce
 }
 
 template <typename ValueType, typename ErrorType>
-const ValueType& expected<ValueType, ErrorType>::value_unchecked() const noexcept
+inline const ValueType& expected<ValueType, ErrorType>::value_unchecked() const noexcept
 {
     return *m_store.template get_at_index<VALUE_INDEX>();
 }
 
+// AXIVION Next Construct AutosarC++19_03-A13.5.2 , AutosarC++19_03-A13.5.3: see doxygen brief section in header
```
Comment (mossmaurice) [inline iceoryx_hoofs/vocabulary/include/iox/detail/expected.inl:232]:
```suggestion
// AXIVION Next Construct AutosarC++19_03-A13.5.2, AutosarC++19_03-A13.5.3: see doxygen brief section in header
```

File: iceoryx_hoofs/vocabulary/include/iox/optional.hpp [hunk 4]
```diff
@@ -228,13 +230,15 @@ class optional final : public FunctionalInterface<optional<T>, T, void>
...
+    // AXIVION DISABLE STYLE AutosarC++19_03-A9.6.1 : False positive. Used type has defined size.
     struct alignas(T) element_t
     {
-        // AXIVION Next Construct AutosarC++19_03-A18.1.1 : safe access is guaranteed since the array
+        // AXIVION Next Construct AutosarC++19_03-A18.1.1, AutosarC++19_03-A1.1.1, AutosarC++19_03-M0.1.3 : required as low level building block, encapsulated in abstraction and not directly used
         // is wrapped inside the optional
         // NOLINTNEXTLINE(hicpp-avoid-c-arrays, cppcoreguidelines-avoid-c-arrays)
         byte_t data[sizeof(T)];
     };
+    // AXIVION Next Construct AutosarC++19_03-A1.1.1 : object size limit is not relevant for containers stored in shared memory
```
Comment (mossmaurice) [inline iceoryx_hoofs/vocabulary/include/iox/optional.hpp:241]: `iox::optional` is not only stored in shared memory, but also on the stack and some users may store them on the heap?

**Review by mossmaurice — COMMENTED** [review]
Only a few smaller things, then we can merge this.

File: iceoryx_hoofs/vocabulary/include/iox/optional.hpp (hunk 4 above)
Comment (FerdinandSpitzschnueffler) [reply iceoryx_hoofs/vocabulary/include/iox/optional.hpp:241]: @mossmaurice To your second question: in https://github.com/eclipse-iceoryx/iceoryx/pull/1749#discussion_r1004575064 we came to the conclusion `UninitializedArray<T, 1> m_data` or implementing an "UninitializedCell" would be overkill for the few remaining occurrences of these warnings.

File: iceoryx_hoofs/functional/include/iox/detail/storable_function.inl (hunk 1 above)
Comment (elBoberido) [reply iceoryx_hoofs/functional/include/iox/detail/storable_function.inl:322]: Well, it's the evaluation order which is now defined :sweat_smile:

File: iceoryx_hoofs/legacy/include/iceoryx_hoofs/cxx/expected.hpp [hunk 5]
```diff
@@ -25,11 +25,11 @@ namespace iox
 /// [[deprecated("Deprecated in 3.0, removed in 4.0, please include 'iox/expected.hpp' instead")]]
 namespace cxx
 {
-/// @deprecated use `iox::error` instead of `iox::cxx::error`
```
[the remark this answers is not in the document]
Comment (elBoberido) [reply iceoryx_hoofs/legacy/include/iceoryx_hoofs/cxx/expected.hpp:28]: That's on my TODO list. Will do that on a follow up PR which also fixes posh.

File: iceoryx_hoofs/vocabulary/include/iox/optional.hpp (hunk 4 above)
Comment (elBoberido) [reply iceoryx_hoofs/vocabulary/include/iox/optional.hpp:241]: I plagiarized from `UninitializedArray`. Have to think of a better reasoning and maybe I remove the suppression for now and fix it in a follow up PR

Comment (elBoberido) [reply iceoryx_hoofs/vocabulary/include/iox/optional.hpp]: Found a better justification :)

**Review by mossmaurice — APPROVED** [review]

File: iceoryx_hoofs/container/include/iox/uninitialized_array.hpp [hunk 6]
```diff
@@ -33,11 +33,13 @@ struct ZeroedBuffer
 {
     struct alignas(ElementType) element_t
     {
-        // AXIVION Next Construct AutosarC++19_03-A18.1.1, AutosarC++19_03-M0.1.3 : required by low level UninitializedArray building block and encapsulated in abstraction, declaration of field in struct for usage elsewhere
```
Comment (FerdinandSpitzschnueffler) [inline iceoryx_hoofs/container/include/iox/uninitialized_array.hpp:36]: Is there no `AutosarC++19_03-A18.1.1` warning anymore?

File: iceoryx_hoofs/container/include/iox/uninitialized_array.hpp [hunk 7]
```diff
@@ -50,11 +52,13 @@ struct NonZeroedBuffer
...
```
Comment (FerdinandSpitzschnueffler) [inline iceoryx_hoofs/container/include/iox/uninitialized_array.hpp:53]: Same here.

File: iceoryx_hoofs/container/include/iox/uninitialized_array.hpp (hunk 6 above)
Comment (elBoberido) [reply iceoryx_hoofs/container/include/iox/uninitialized_array.hpp:36]: It seems the scan is unreliable :/

File: iceoryx_hoofs/functional/include/iox/detail/storable_function.inl (hunk 1 above)
Comment (elBoberido) [reply iceoryx_hoofs/functional/include/iox/detail/storable_function.inl]: Actually it was just an unreliable scan :/

File: iceoryx_hoofs/container/include/iox/uninitialized_array.hpp (hunk 7 above)
Comment (elBoberido) [reply iceoryx_hoofs/container/include/iox/uninitialized_array.hpp:53]: fxied

**Comment by elBoberido** [comment]
@FerdinandSpitzschnueffler @mossmaurice If you wish, I can squash the `Fix more` commits after you did the review

**Review by FerdinandSpitzschnueffler — DISMISSED** [review]

**Comment by FerdinandSpitzschnueffler** [comment]
> @FerdinandSpitzschnueffler @mossmaurice If you wish, I can squash the `Fix more` commits after you did the review

If you have to change something or rebase, please squash those commits.

**Comment by mossmaurice** [comment]
@elBoberido LGTM, thanks for your patience!

> If you wish, I can squash the Fix more commits after you did the review

Yes, please squash them.

**Comment by elBoberido** [comment]
@mossmaurice @FerdinandSpitzschnueffler done. I also needed to fix a merge conflict with the new `@deterministic` flag in `UninitializedArray`. Except that nothing changed.

**Review by FerdinandSpitzschnueffler — DISMISSED** [review]

**Comment by elBoberido** [comment]
@mossmaurice @FerdinandSpitzschnueffler another rebase to incorporate the fix for the integration test setup

**Review by FerdinandSpitzschnueffler — APPROVED** [review]

File: iceoryx_hoofs/memory/include/iox/detail/scoped_static.inl [hunk 8]
```diff
@@ -25,7 +25,7 @@ template <typename T, typename... CTorArgs>
 inline ScopeGuard makeScopedStatic(T& memory, CTorArgs&&... ctorArgs) noexcept
 {
     memory.emplace(std::forward<CTorArgs>(ctorArgs)...);
-    return ScopeGuard([&memory]() { memory.reset(); });
+    return ScopeGuard([&memory]() noexcept { memory.reset(); });
```
Comment (mossmaurice) [inline iceoryx_hoofs/memory/include/iox/detail/scoped_static.inl:28]: Well, this is only `noexcept` if the underlying type stored does not throw. You could add a todo here, to derive the exception specifier from `T`.

**Review by mossmaurice — APPROVED** [review]

File: iceoryx_hoofs/memory/include/iox/detail/scoped_static.inl (hunk 8 above)
Comment (elBoberido) [reply iceoryx_hoofs/memory/include/iox/detail/scoped_static.inl:28]: Something for the future :)
Will add the todo in the next PR

**Pull request merged**